### ISSUE TITLE
Extraction fails for a metaprogram with a spec

### ISSUE THREAD

**Issue by mtzguido** [FStarLang/FStar#1479]
If I add a spec to the `fail` primitive, then this code:
```f#
let skip_guard () : Tac unit =
    if is_guard ()
    then smt ()
    else fail ""
```
Fails extraction with:
```
FStar.Tactics.Effect.fst(26,20-26,41): (Error 76) Ill-typed application: application is ((match a with
            | true  -> (reify (smt ()))
            |_  -> (reify (fail ""))) (decr_depth (set_proofstate_range ps' (prims_to_fstar_range FStar.Tactics.Derived.fst(130,4-132,16)))))
remaining args are (decr_depth (set_proofstate_range ps' (prims_to_fstar_range FStar.Tactics.Derived.fst(130,4-132,16))))
ml type of head is Obj.t
```
Others work fine it seems. See: ae448aa800cd2176fd2fac1b3ef952cf4f5f5a4d

**Comment by mtzguido** [comment]
The WP parts of the OCaml types for each branch differ. Here's a workaround
```
let skip_guard () : Tac unit =
    if is_guard ()
    then smt () <: TAC unit (fun _ _ -> False)
    else fail ""
```
Looking into a fix

**Comment by victor-dumitrescu** [comment]
I've been looking into this but I'm not yet entirely sure what the cause is. In extraction (`FStar.Extraction.ML.Term.check_term_as_mlexpr`), the body of the match in `skip_guard` is expected to have type `Prims.bool  ->  (unit, unit) FStar_Tactics_Effect._dm4f_TAC_repr`. This is the case when there's no spec, but when there is, its type is `Prims.bool  ->  (unit, Obj.t) FStar_Tactics_Effect._dm4f_TAC_repr`, which triggers a coercion, which turns the whole application into an `Obj.t`, hence the ill-typed application error.

**Comment by mtzguido** [comment]
I think we're reifying the computation before translating, getting a `_DM4F_TAC_repr a wp`, and translating `Tm_let/Tm_match` as a type just returns `Obj.t`. Maybe `translate_term_to_mlty` could do better, but who cares? Why not nuke all effect arguments into `Prims.unit`? I just tried that and it seems to work, but running CI to be sure. I suppose it might generate some ill-typed **if** we had indexed DM4F effects, but we don't.
